ksql-engine/src/main/java/io/confluent/ksql/function/udaf/map/HistogramUdaf.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.confluent.ksql.function.udaf.map;
+
+import com.google.common.collect.Maps;
+import io.confluent.ksql.function.udaf.TableUdaf;
+import io.confluent.ksql.function.udaf.UdafDescription;
+import io.confluent.ksql.function.udaf.UdafFactory;
+import java.util.Map;
+
+@UdafDescription(name = "histogram", description = "Returns a map of each distinct value from the"
+    + " input Stream or Table and how many times each occurs."
+    + " \nNot applicable for complex types (map, array, or struct)."
+    + " \nThis version limits the size of the resultant Map to 1000 entries. Any entries added"
+    + " beyond this limit will be ignored.")
+public final class HistogramUdaf {
+
+  private static final int LIMIT = 1000;
+  
+  private HistogramUdaf() {
+  }
+  
+  private static <T> TableUdaf<T, Map<T, Long>> histogram() {
+    return new TableUdaf<T, Map<T, Long>>() {
```

Comment:
@apurvam raised a good point on the key for the histogram. We assume that KSQL map type keys are `STRING` type. Here we can have map with different types depending on the column type we pass to the function.
One option is to cast the column value to string and use string value for the map key here.
This will make the result map consistent with the KSQL map type.

Reply:
yes, @apurvam makes a very good point! I've been testing it again this morning and, interestingly enough, it does "what you would expect" - i.e. it creates a MAP<your-input-type, long> and writes it out correctly but unfortunately the resultant schema describes it as a MAP<string, long>, like we do for all maps.
I see 2 short-term options, until maps allow other primitive key types:
1 - toString all the input values, so that we always return a map<stringified-value, long>, regardless of input type.
2 - simply comment or remove the non-string signatures so you can't actually run this on non-string input values. 
I think i prefer #2 because it means there's no breaking change when we loosen the restriction on map keys later. If a user wants the behavior of #1 i guess they can still get it by CASTing the values to string before counting them, it's just a little clumsy.
WDYT ?

Reply:
I like option 2. Better to do less implicit conversions.

Reply:
I'll defer to @hjafarpour on this though :)

Reply:
ok, option 2 it is.